docs/story.rst.j2

```diff
@@ -0,0 +1,153 @@
+        fragment.
+#}
+{% macro emit_docs_link(link) %}
+{% set matched = link.target | match('^/docs/(.+?)\\.rst(?:#([a-z\-]+))?$') %}
+* {{ printable_relation(link) }} :ref:`{{ matched.group(1) }}{# {% if matched.group(2) %}:{{ matched.group(2).replace('-', ' ') }}{% endif %} #}`
+{% endmacro %}
+
+{% if INCLUDE_TITLE %}
+{% set depth = STORY.name | findall('/') | length - 1 %}
+{% set title_underline = '=~^:-><'[depth] %}
+{% if STORY.title and STORY.title != STORY.node.parent.get('title') %}
+    {% set title = STORY.title %}
+{% else %}
+    {% set title = STORY.name | regex_replace('.*/', '') %}
+{% endif %}
+
+.. _{{ STORY.name | strip }}:
+
+{{ title | strip }}
+{{ title_underline * title | length }}
+{% endif %}
+
+{# Summary, story and description #}
+{% if STORY.summary and STORY.summary != STORY.node.parent.get('summary') %}
+{{ STORY.summary | strip }}
+{% endif %}
+
+{% if STORY.story != STORY.node.parent.get('story') %}
+*{{ STORY.story | strip }}*
+{% endif %}
+
+{# Insert note about unimplemented feature (leaf nodes only) #}
+{% if not STORY.node.children and not STORY.implemented %}
+.. note:: This is a draft, the story is not implemented yet.
+{% endif %}
+
+{% if STORY.description and STORY.description != STORY.node.parent.get('description') %}
+{{ STORY.description }}
+{% endif %}
+
+{# Examples #}
+{% if STORY.example and STORY.example != STORY.node.parent.get('example') %}
+    {% for example in STORY.example %}
+        {% if example == STORY.example | first %}
+**Examples:**
+        {% endif %}
+
+{{ render_example(example) }}
+    {% endfor %}
+{% endif %}
+
+{# Status #}
+{% if not STORY.node.children %}
+    {% if STORY.status %}
+**Status:** {{ STORY.status | listed }}
+    {% else %}
+**Status:** idea
+    {% endif %}
+{% endif %}
+
+{# Links #}
+{% for link in STORY.link.get() %}
+{# Links pointing to docs #}
+{% if link.target | match('^/docs/') %}
+{{ emit_docs_link(link) }}
+
+{# Links pointing to step core implementation #}
+{% elif link.target | match('^/tmt/steps(?:/__init__.py)?$') %}
+{{ emit_tmt_link(link, '/tmt/(steps)(?:/__init__.py)?') }}
+
+{# Links pointing to step implementations #}
+{% elif link.target | match('^/tmt/steps/([a-z_]+)(?:$|(?:/__init__.py))$') %}
+{{ emit_tmt_link(link, '/tmt/steps/([a-z_]+)(?:$|(?:/__init__.py))') }}
+
+{# Links pointing to step plugins #}
+{% elif link.target | match('^/tmt/steps/([a-z_]+/[a-z_]+)\\.py$') %}
+{{ emit_tmt_link(link, '/tmt/steps/([a-z_]+/[a-z_]+)\\.py') }}
+
+{# Links pointing to other tmt code #}
+{% elif link.target.startswith('/tmt') %}
+{{ emit_tmt_link(link, '/(.*)') }}
+
+{# Links pointing to stories #}
+{% elif link.target | match('^/stories') %}
+* {{ printable_relation(link) }} :ref:`{{ link.target }}`
+
+{# Links pointing to tests #}
+{% elif link.target | match ('^/tests') %}
+{{ emit_tmt_link(link, '(.*)') }}
+
+{# Links pointing to plans #}
+{% elif link.target | match ('^/plans') %}
+{{ emit_tmt_link(link, '(.*)') }}
```

Comment:
Links for `tests` and `plans` will not work for those which are using virtual hierarchy, for example:

    - relates: /plans/install/pip/mini

This would point [here](https://github.com/teemtee/tmt/tree/main//plans/install/pip/mini). Perhaps we could use the `web_link()` method to generate the link instead?

Reply:
Good point. What is the canonical way of getting an object whose `web_link()` I could call? The template is rendering a story, then a link, but the link's target is a string and I need to get my hands on `Test`, `Plan` or `Story` target points. I don't see a single method in `Tree` I could use, but I guess this is doable by checking what the first part of the target is, `test` or `plan`, and then calling the corresponding method, `Tree.tests(names=[link.target])`, for example.

Any shortcuts, or better ideas? This might be quite clumsy.

Reply:
> calling the corresponding method, `Tree.tests(names=[link.target])`, for example

Yeah, that was the first thing which came to my mind as well. And I don't think there is a more straightforward way. What about implementing a `tmt` counterpart of `fmf.Tree.find()` which finds a single node by name and returns a `Tree` object? In `tmt` case it would take care of picking/creating `Test`, `Plan` or `Story` object as well.

Reply:
Improved in a09ce0adc6c6fdf64570814ace097c9d0f1e6a24

Reply:
Thanks! Now works nicely. Although I have to admit that the jinja jungle is becoming little bit hard to read/understand.